### PR TITLE
add a role to manage products, repositories, and repository_sets

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

**Comment by mdellweg** [comment]
Where do you want to go with this? ATM it's a role that calls a single module once without even mangling any parameter, and i do not see any benefit.
What about a role that accepts a list of products that contain lists of repositories with URLs. It then `state=present`s them and syncs all (maybe in parallel)?

**Comment by ehelms** [comment]
I agree @mdellweg and this presents the first opportunity to discuss encouraging a best practices structure for how a user might define and manage their data. A format we've used to do this using RHEL as an example:

```
rhel_product:
  name: Red Hat Enterprise Linux Server
  repository_sets:
    - name: Red Hat Enterprise Linux 7 Server (RPMs)
      basearch: x86_64
      releasever: 7Server
    - name: Red Hat Enterprise Linux 6 Server (RPMs)
      basearch: x86_64
      releasever: 6Server
    - name: Red Hat Enterprise Linux 5 Server (RPMs)
      basearch: x86_64
      releasever: 5Server
    - name: Red Hat Enterprise Linux 7 Server - Extras (RPMs)
      basearch: x86_64
      releasever: ''
    - name: Red Hat Enterprise Linux 7 Server - Optional (RPMs)
      basearch: x86_64
      releasever: ''
  repositories:
    "7Server":
      name: Red Hat Enterprise Linux 7 Server RPMs x86_64 7Server
    "6Server":
      name: Red Hat Enterprise Linux 6 Server RPMs x86_64 6Server
    "5Server":
      name: Red Hat Enterprise Linux 5 Server RPMs x86_64 5Server
    "7Server Extras":
      name: Red Hat Enterprise Linux 7 Server - Extras RPMs x86_64
    "7Server Optional":
      name: Red Hat Enterprise Linux 7 Server - Optional RPMs x86_64 7Server
```

**Comment by wbclark** [comment]
I've updated this role to enable and synchronize repositories, using the below structure for the data:

```yaml
- hosts: localhost
  roles:
    - role: theforeman.foreman.reposync
      vars:
        server_url: https://foreman.example.com
        username: "admin"
        password: "changeme"
        organization: "Default Organization"
        products:
          - name: Red Hat Enterprise Linux Server
            repository_sets:
              - name: Red Hat Enterprise Linux 7 Server (RPMs)
                basearch: x86_64
                releasever: 7Server
              - name: Red Hat Enterprise Linux 6 Server (RPMs)
                basearch: x86_64
                releasever: 6Server
              - name: Red Hat Enterprise Linux 7 Server - Extras (RPMs)
                basearch: x86_64
                releasever: ''
              - name: Red Hat Enterprise Linux 7 Server - Optional (RPMs)
                basearch: x86_64
                releasever: ''
          - name: Red Hat Enterprise Linux for x86_64
            repository_sets:
              - name: Red Hat Enterprise Linux 8 for x86_64 - BaseOS (RPMs)
                basearch: x86_64
                releasever: 8Server
              - name: Red Hat Enterprise Linux 8 for x86_64 - AppStream (RPMs)
                basearch: x86_64
                releasever: 8server
```

Note that it was sufficient to define `repository_sets` per `product` and not yet necessary to define the actual `repositories`. This is because the `repository_sync` module synchronizes all repositories in a product by default when specific repositories are not passed.

TODO: Write and record tests, although I expect I'll face a similar issue to what I'm seeing with #978 , which is `"msg": "Error while performing upload on subscriptions: Can't overwrite existing cassette ('fixtures/manifest_role-0.yml') in your current record mode ('none')`

**Comment by wbclark** [comment]
At this time the role is working to enable repositories and trigger sync tasks.

However I'm getting errors while syncing on my `centos7-katello-nightly` box so I'm going to provision a more stable release to record the tests.

File: roles/repositories/defaults/main.yml
Comment (evgeni) [inline roles/repositories/defaults/main.yml]: this is the default
```suggestion
```

File: roles/repositories/tasks/main.yml
Comment (evgeni) [inline roles/repositories/tasks/main.yml]:
```suggestion
    validate_certs: "{{ validate_certs | default(omit) }}"
```

File: roles/repositories/tasks/main.yml
Comment (evgeni) [inline roles/repositories/tasks/main.yml]:
```suggestion
    validate_certs: "{{ validate_certs | default(omit) }}"
```

File: tests/test_playbooks/repositories_role.yml
Comment (evgeni) [inline tests/test_playbooks/repositories_role.yml:28]: this needs to be created first against `localhost`

File: tests/test_playbooks/repositories_role.yml
Comment (evgeni) [inline tests/test_playbooks/repositories_role.yml]:
```suggestion
        validate_certs: "{{ foreman_validate_certs }}"
```

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]: this role is called `repositories` ;)

File: roles/repositories/tasks/main.yml
Comment (evgeni) [inline roles/repositories/tasks/main.yml:13]: What if I want to enable multiple repo-sets of the same product? I think currently this will fail (or rather the last repo-set will win)

**Comment by evgeni** [comment]
I have one major design issue with this role: It's called `repositories`, while in reality it's only `redhat_(cdn_)repositories`. Nothing wrong with providing a Red Hat specific role here, but I think we should *try* to also accommodate other users as good as possible.

My vision would be that the user has a data structure (call it `products`) that encapsulates all required knowledge to obtain these products (and can later be used to provide them to the clients). Based on the example playbook in your readme, it could be something like this:

```yaml
products:
  - name: Red Hat Enterprise Linux Server
    type: redhat
    repository_sets:
      - name: Red Hat Enterprise Linux 7 Server (RPMs)
        basearch: x86_64
        releasever: 7Server
  - name: CentOS 8
    type: yum
    repositories:
      - name: BaseOS x86_64
        url: http://mirror.centos.org/centos/8/BaseOS/x86_64/os/
  - name: Debian 10
    type: deb
    repositories:
      - name: main
        url: http://deb.debian.org/debian
        deb_components: main
        deb_architectures: amd64
        deb_releases: buster
```

The user feeds that list to a `repositories` role, which invokes a `redhat_repositories` (what you have written), `yum_repositories` and `deb_repositories` (both to be written, they create products and repositories the "normal way") roles, based on the `type` and finishes off with a `repository_sync` for all products.

(the alternative is not to have `type` and make the user invoke the individual roles, but seems less friendly)

**Comment by mdellweg** [comment]
I think, the type is really something on the repository level *. Also why split it into different roles. We can use that structure and just one `repositories` role to set it all up.

*) IMHO it can be beneficial to have the same (software-)product packaged for different OSes in the same (katello-)product.

**Comment by ehelms** [comment]
> I think, the type is really something on the repository level *.

Agreed, I think it might create overhead for users to have to define type at the product level when it only applies to the repository itself (and a product can have multiple types in it).

**Comment by evgeni** [comment]
The problem of doing that on the repository level is that in the Red Hat case, you don't ever create products (or repositories), Katello does all that for you.

**Comment by ehelms** [comment]
This is where I would love to see either `vars.yaml.example` or `vars/products.yaml.example` or something equivalent that lays out an example layout of the products structure at the top level for users to reference or copy.

**Comment by ehelms** [comment]
> The problem of doing that on the repository level is that in the Red Hat case, you don't ever create products (or repositories), Katello does all that for you.

In that case, I would think the types, if at the product level, would be `redhat` and `custom`. But I wonder if we can omit that and let the existence of `repository_sets` and `repositories` do the switching for us given that is the power of roles?

**Comment by evgeni** [comment]
Ah, yeah, that'd be possible. Great idea!

**Comment by mdellweg** [comment]
> given that is the power of roles?

I couldn't have said it better.

**Comment by evgeni** [comment]
So, to reiterate, the new structure would be:

```yaml
products:
  - name: Red Hat Enterprise Linux Server
    repository_sets:
      - name: Red Hat Enterprise Linux 7 Server (RPMs)
        basearch: x86_64
        releasever: 7Server
  - name: CentOS 8
    repositories:
      - name: BaseOS x86_64
        content_type: yum
        url: http://mirror.centos.org/centos/8/BaseOS/x86_64/os/
  - name: Debian 10
    repositories:
      - name: main
        content_type: deb
        url: http://deb.debian.org/debian
        deb_components: main
        deb_architectures: amd64
        deb_releases: buster
  - name: Foreman Client
    repositories:
      - name: Foreman Client Debian
        url: http://deb.atix.de/debian
        content_type: deb
        deb_components: main
        deb_architectures: amd64
        deb_releases: buster
      - name: Foreman Client CentOS
        url: http://yum.theforeman.org/client/latest
        content_type: yum
```

Or is content_type not required either?

**Comment by ehelms** [comment]
> Or is content_type not required either?

Definitely required on a repository.

**Comment by ehelms** [comment]
I am torn if I want a role that enables/ensures products and repositories exist *and* sync's the repositories at the same time. I definitely want a role that I can use to sync repositories which would then create some duplication with this role. Further, if while managing my repositories I want to make a change to a repository I may not want to sync it (e.g. if I were to enable HTTP on the repository). However, if I changed the URL of the repository, in that case I may want to sync? But perhaps we should give user more control over that via two separate roles:

 - repositories
 - repository_sync

**Comment by bagasse** [comment]
Sorry to jump in this conversation :smile:. If we manage also custom products/repos, can we manage also some other "content management" things that products/repos depends on, like content_credential, sync_plans ? We worked on some playbooks originally, that we are moving to collection/roles that follow some usecases we have (like I described in [this comment](https://github.com/theforeman/foreman-ansible-modules/issues/646#issuecomment-608480355)). Has stated here our primary goal is to define some data structures that the users follow and just provide it as variables to role(s) that enforce it in our katello instances.
Maybe we can only sync repos only on change by default (and control this behavior via a variable) ?

**Comment by evgeni** [comment]
@ehelms I'd prefer not to have logic that decides whether a sync is needed or not based on what has changed (url vs http publish in your example), this feels too magic. A tristate "sync: always/change/never" might be a fine compromise. Or just have two roles that accept the same data (structure) for management and syncing.

@bagasse heey, welcome back! Ideally, yes, we should also allow management of the supplementary entities, the question is always how much in detail can we go without making it too complicated to use. Take the sync plans for example, we could add a sync_plan key (with a cron like value) to each product or repo in the example above and autogenerate sync  plan entries from that, or we could define plans at the same level as the products and the plans refer to products. As a user I'd probably prefer the first? But then it again becomes magic in a way...

**Comment by bagasse** [comment]
> As a user I'd probably prefer the first? But then it again becomes magic in a way...

I don't know for others, but we don't use too much sync plans, we just have one for internal repo sync (hourly) and two other (daily) for external repos (centos, epel, ovirt, foreman/katello...), so defining sync_plans at the same level as products and refer to it in products is fine for us.

File: roles/repositories/tasks/main.yml
Comment (ehelms) [inline roles/repositories/tasks/main.yml]: I'm still not sold that enabling repositories should imply a sync of the repositories.

File: roles/repositories/tasks/main.yml
Comment (ehelms) [inline roles/repositories/tasks/main.yml]: Is there a benefit to putting this inside the repositories role instead of split out into it's own role? I would think the latter provides more composibility for users.

File: roles/repositories/README.md
Comment (ehelms) [inline roles/repositories/README.md]:
```suggestion
This role enables Red Hat Repositories, creates Custom Products and Repositories, and defines Sync Plans.
```

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml:52]: The benefit is that defining it in the same role allows us to specify that a repository should be synced only when changed.

I'm still debating whether that's worth doing, as well as researching the data structures involved when registering the results of the with_subelements loops above.

**Comment by wbclark** [comment]
I want to extend a big thank you to @ehelms @mdellweg @evgeni @bagasse for all the reviews. I have greatly benefited from the effort you've put towards sharing your thoughts and requirements for this feature.

I've pushed an update and want to summarize what has changed:

I. The role now supports creating Products and custom Repositories.
II. The role now allows to define Sync Plans and assign products to them.
III. All pieces of the data structure are optional. You may or may not enable Red Hat Repositories. You may or may not create custom Products and Repositories. You may or may not define Sync Plans. If any of these are empty, e.g. `Products` is passed as an empty list, the role should not fail.

I still haven't touched the synchronize repository task and am leaning towards defining a tri-state initial sync strategy at the Product level that allows syncs to be performed always, never, or only when the Product is changed.

I need to update the tests as well and will get to work on that shortly.

I'm still interested in any further feedback, and I want to solicit feedback in particular about the clarity of the README. Any further feedback is highly valued.

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml]: Ah, I thought this comment was relevent to Syncs, not Sync Plans. My apologies. Please disregard my prior comment.

Comment (wbclark) [reply roles/repositories/tasks/main.yml]: The benefit here is that the user can define data for their Products alongside data for their Sync Plans.

Regarding composibility, I've taken care to ensure that each piece should be optional and tasks should be skipped when the required data structure is not defined.

@ehelms do you think there would be value in defining this role to be re-usable so that it can be passed multiple times with different variables for any users who desire that approach?

Comment (ehelms) [reply roles/repositories/tasks/main.yml]: I'm just having a hard time seeing the value provided by putting this all in a single role versus providing users more flexibility by having roles for configuring repositories, sync plans and taking the action to sync a repository. I think `sync` is a great example of this. Not all users will want to sync their repositories when they create or enable them. Users will want to sync repositories out of band of enable and creation. Therefore, we will end up with a stand-alone sync role anyway. So I perceive it as better to just write the stand-alone sync role (it could always be included by the repositories role).

Comment (evgeni) [reply roles/repositories/tasks/main.yml]: I think at some point I expressed the desire of having a "meta" data structure that defines "all things content" and all content related roles feed from the same structure, to avoid duplication of information in similar but not identical structures. This might be influenced by that idea?

Comment (wbclark) [reply roles/repositories/tasks/main.yml]: Yes. One goal I have is that the user should never have to define the same data in multiple places.

That's still possible with when composing role, but what it looks like is a single composite role, which receives the data once, and passes it to the child roles.

IMO at that point it's easier and more maintainable to just have a single role

Comment (ehelms) [reply roles/repositories/tasks/main.yml]: > That's still possible with when composing role, but what it looks like is a single composite role, which receives the data once, and passes it to the child roles.

I do not really understand this nuance. But I do not feel my point was addressed either. A dedicated sync role gives more power to users and allows use in this case.

A user will definitely have a common data structure spanning these roles. The repositories structure might be used by a user when defining their content views for example.

I think we are in that weird space between providing "workflows" and providing "components" for use by a user. My concern is if a user does not want to initiate a sync everytime they change something (add a repository, or product, change a property of a repository) they have to incur a sync. And maybe they prefer to let sync plans drive that, maybe their CI/CD but the control is taken away from them. Whereas a stand alone sync role allows the user to construct playbooks that define their workflow.

File: roles/repositories/README.md
Comment (mdellweg) [inline roles/repositories/README.md]: Custom repositories require a name, too, right?

Personal preference: Include a line break after each sentence.

File: roles/repositories/README.md
Comment (mdellweg) [inline roles/repositories/README.md]: Maybe make the "points" a bullet list.

File: roles/repositories/README.md
Comment (mdellweg) [inline roles/repositories/README.md]:
```suggestion
              - Red Hat Enterprise Linux Server
              - Red Hat Enterprise Linux for x86_64
```

File: roles/repositories/tasks/main.yml
Comment (mdellweg) [inline roles/repositories/tasks/main.yml]: Any reason, we are overriding katello's default here?

File: roles/repositories/tasks/main.yml
Comment (evgeni) [reply roles/repositories/tasks/main.yml]: After thinking more on this, I agree, sync plans should go into an own role.

File: roles/repositories/tasks/main.yml
Comment (evgeni) [reply roles/repositories/tasks/main.yml]: I think, at some point, we mentioned a flag that will result in a sync "always", "never", "if changed".

Honestly, this could also be done with an own role (and give the user better flexibility what they sync). The only caveat is that `import_role` doesn't register a result, so you can't automatically see if something changed by running the role. The role would need to register that itself and we'd have to document that for the user.

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml:13]: This has not been the case in my testing. For example the test playbook here has 2 repository sets under the same product, and both are successfully enabled.

I admit I am very impressed by Ansible's `with_subelements`

File: tests/test_playbooks/repositories_role.yml
Comment (wbclark) [reply tests/test_playbooks/repositories_role.yml:28]: Done.

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml]: No and indeed it fails for Debian repositories unless I use `default(omit)` so I have made the change.

File: roles/repositories/README.md
Comment (wbclark) [reply roles/repositories/README.md]: I've made a first pass at cleaning that up

File: roles/repositories/README.md
Comment (wbclark) [reply roles/repositories/README.md]: Done, but I admit I'm still not fully satisfied with the overall flow and formatting of the doc

**Comment by wbclark** [comment]
I have split all functionality related to sync plans off with https://github.com/theforeman/foreman-ansible-modules/pull/1065

File: roles/repositories/README.md
Comment (ehelms) [inline roles/repositories/README.md]: Nit: `defines Products, enabled Red Hat Repositories and Custom Repositories.`

File: roles/repositories/README.md
Comment (ehelms) [inline roles/repositories/README.md:54]: Mind adding a second repostitory to this example for AppStream to show why you would want to do a `CentOS 8` product instead of a `CentOS` product with 7 and 8 in it?

**Comment by ehelms** [comment]
I just had a few thoughts around the README since I know users will look to those examples as a best practice. Looks good @wbclark !

File: roles/repositories/tasks/main.yml
Comment (wbclark) [inline roles/repositories/tasks/main.yml:44]: TODO: Should technically be able to define a Product without defining Repositories on that Product. I'm not sure if there is a real use for that, but Katello allows it and I noticed it would be nice to have when developing test playbooks for other roles.

File: roles/repositories/README.md
Comment (evgeni) [reply roles/repositories/README.md]: Should this rather be "This role defines Products and Custom Repositories and enables Red Hat Repositories."
As it does not *enable* custom repos?

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]:
```suggestion
- `products`: List of products to manage.
```

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md:68]:
```suggestion
        url: https://yum.theforeman.org/client/latest/
```

File: roles/repositories/README.md
Comment (wbclark) [reply roles/repositories/README.md]: Yes, I think so. I'm also leaning to renaming the role `Products` with the shape it is taking

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]:
```suggestion
      - name: Foreman Client Debian 10
        url: https://apt.atix.de/Debian10/
        content_type: deb
        deb_components: main
        deb_architectures: amd64
        deb_releases: stable
```

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]:
```suggestion
      - name: Foreman Client CentOS 7
        url: https://yum.theforeman.org/client/latest/el7/x86_64/
```

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]:
```suggestion
      - name: Debian 10 main
```

File: roles/repositories/tasks/main.yml
Comment (evgeni) [reply roles/repositories/tasks/main.yml:13]: Cool,thanks!

File: roles/repositories/tasks/main.yml
Comment (evgeni) [inline roles/repositories/tasks/main.yml:13]: the repo set module has a `all_repositories` parameter: https://theforeman.github.io/foreman-ansible-modules/develop/plugins/repository_set_module.html#parameter-all_repositories

Should we expose that here? There is also an example how to combine that with `repositories` so that the user can have a single data structure but some sets have `repositories` and the others end up being `all`.

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md:1]: I wonder if `theforeman.foreman.content` is not a more appropriate name? but wouldn't block on this

File: roles/repositories/tasks/main.yml
Comment (evgeni) [inline roles/repositories/tasks/main.yml:47]: would you mind exposing the "other" params of the module here too, with `|default(omit)`, see https://theforeman.github.io/foreman-ansible-modules/develop/plugins/repository_module.html

this would allow people also manage container/ansible/puppet repos with that

and we could document something like "the `repositrory` takes any parameter the `repository` module supports"

File: roles/repositories/README.md
Comment (ehelms) [reply roles/repositories/README.md:1]: Content seems so broad spectrum that given this is focused on repositories, I would stick to that personally.

Comment (evgeni) [reply roles/repositories/README.md:1]: wfm, was just a thought :)

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml:47]: Done.

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml:13]: I added a new task for Products which are specified with `all_repositories: True` and documented it appropriately. I added data in the test playbook to test for this task as well.

After using the module in this way, my general feeling is that this is something we should discourage users from doing unless they fully understand the consequences of enabling a dozen point releases that they are likely not going to use.

File: roles/repositories/README.md
Comment (wbclark) [reply roles/repositories/README.md:54]: Done.

File: roles/repositories/tasks/main.yml
Comment (wbclark) [reply roles/repositories/tasks/main.yml:13]: A caveat is that `all_repositories` requires either `name` + `product`, OR `label` to be specified. I added support for the labels but not for product + name. I could add that still if there is value in it (the `name`s are more specific I think and avoid the problem of enabling a dozen unused point releases?) but it would require a 3rd task or a bit of rethinking my approach to the looping.

EDIT: the reason for the awkwardness is that you would need to specify the `repository_sets` in order to give them `name`s, which breaks assumptions in the prior `with_subelements` loop over Products with repository sets. We could instead expect it to be called something other than repository_sets in this case but that makes the data structure itself awkward.

Comment (wbclark) [reply roles/repositories/tasks/main.yml:13]: I also tried to test multiple labels but using the `rhel-server-rhscl-6-rpms` label resulted in working Ansible but "The path /content/dist/rhel/server/6/6.9/i386/rhscl/1/os does not seem to be a valid repository. If you think this is an error, please try refreshing your manifest." from Katello

**Review by ehelms — APPROVED** [review]
Everything looks good to me, I'm excited to get this particular role out there for use. Will let @evgeni have final say in review.

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md]:
```suggestion
This role defines Products and Custom Repositories and enables Red Hat Repositories.
```

File: roles/repositories/README.md
Comment (evgeni) [inline roles/repositories/README.md:14]: Why the `label` limitation? Can't it be enabled with `product` and `name` too?

Comment (evgeni) [reply roles/repositories/README.md:14]: Ah, I see in https://github.com/theforeman/foreman-ansible-modules/pull/1027/files#r541297185

Okay for now I guess.

**Review by evgeni — CHANGES_REQUESTED** [review]
one doc nitpick ;)

File: roles/repositories/README.md
Comment (wbclark) [reply roles/repositories/README.md]: Yeah, this feels a lot more natural. :+1:

**Review by evgeni — APPROVED** [review]

**Comment by evgeni** [comment]
@wbclark last thing, really the last (:sweat_smile:) can you please add a changelog entry? new modules get that automatically, but new roles do not.

**Comment by evgeni** [comment]
I don't think `roles` is a valid section for changelogs, can you please use `minor_changes`?

**Review by evgeni — APPROVED** [review]

**Pull request merged**
